src/hookable.ts

```diff
@@ -113,15 +117,29 @@ export class Hookable <
   }
 
   callHook<NameT extends HookNameT> (name: NameT, ...args: Parameters<InferCallback<HooksT, NameT>>) {
-    return serialCaller(this._hooks[name] || [], args)
+    return this.callHookWith(serialCaller, name, ...args)
   }
 
   callHookParallel<NameT extends HookNameT> (name: NameT, ...args: Parameters<InferCallback<HooksT, NameT>>) {
-    return parallelCaller(this._hooks[name] || [], args)
+    return this.callHookWith(parallelCaller, name, ...args)
   }
 
   callHookWith<NameT extends HookNameT, CallFunction extends (hooks: HookCallback[], args: Parameters<InferCallback<HooksT, NameT>>) => any> (caller: CallFunction, name: NameT, ...args: Parameters<InferCallback<HooksT, NameT>>): void | ReturnType<CallFunction> {
-    return caller(this._hooks[name] || [], args)
+    parallelCaller(this._before || [], [name, ...args])
+    const result = caller(this._hooks[name] || [], args)
+    if (result as any instanceof Promise) {
+      return result.finally(() => parallelCaller(this._after || [], [name, ...args]))
+    }
+    parallelCaller(this._after || [], [name, ...args])
+    return result
+  }
+
+  beforeHook (fn: (name: string, ...args: any[]) => any) {
```

Comment:
Since hooks are typed, is there any hope we can also type these two by inferring from them?

Reply:
Also hooks shouldn't return anything

Reply:
It was meant to allow for promises, but I can be more explicit.

Reply:
I would keep before/after non async because their main purpose is being for interceptors